pkg/infradb/utils.go

```diff
@@ -0,0 +1,294 @@
+		if preSub.Name == componentName {
+			preSubscriber = preSub
+			break
+		}
+	}
+
+	if preSubscriber == nil {
+		deferErr = fmt.Errorf("no pre-replay subscriber for %s", componentName)
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	// Notify the preReplay subscriber
+	actionData := actionbus.NewActionData()
+	deferErr = actionbus.ABus.Publish(actionData, preSubscriber)
+	if deferErr != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	// Waiting for the pre-replay procedure to finish
+	deferErr = <-actionData.ErrCh
+	close(actionData.ErrCh)
+
+	if deferErr != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	log.Printf("startReplayProcedure(): Component %s has successfully executed pre-replay steps", componentName)
+
+	objectTypesToReplay := getObjectTypesToReplay(componentName)
+
+	objectsToReplay, subsForReplay, err := gatherObjectsAndSubsToReplay(componentName, objectTypesToReplay)
+
+	if err != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", err)
+		deferErr = err
+		return
+	}
+
+	// Releasing the lock as all the operations in the DB has finished
+	globalLock.Unlock()
+
+	// Notify task manager to continue processing tasks as
+	// the replay of objects in the DB has finished
+	taskmanager.TaskMan.ReplayFinished()
+
+	createReplayTasks(objectsToReplay, subsForReplay)
+}
+
+// getObjectTypesToReplay collects all the types of object to be replayed
+// which are related to the component that called the replay.
+func getObjectTypesToReplay(componentName string) []string {
+	objectTypesToReplay := []string{}
+
+	bpSubs := eventbus.EBus.GetSubscribers("bridge-port")
+	sviSubs := eventbus.EBus.GetSubscribers("svi")
+	lbSubs := eventbus.EBus.GetSubscribers("logical-bridge")
+	vrfSubs := eventbus.EBus.GetSubscribers("vrf")
+
+	for _, vrfSub := range vrfSubs {
+		if vrfSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "vrf")
+			break
+		}
+	}
+
+	for _, lbSub := range lbSubs {
+		if lbSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "logical-bridge")
+			break
+		}
+	}
+
+	for _, sviSub := range sviSubs {
+		if sviSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "svi")
+			break
+		}
+	}
+
+	for _, bpSub := range bpSubs {
+		if bpSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "bridge-port")
+			break
+		}
+	}
+
+	return objectTypesToReplay
+}
+
+// nolint: funlen, gocognit
+func gatherObjectsAndSubsToReplay(componentName string, objectTypesToReplay []string) ([]interface{}, [][]*eventbus.Subscriber, error) {
+	objectsToReplay := []interface{}{}
+	subsForReplay := [][]*eventbus.Subscriber{}
+
+	bpSubs := eventbus.EBus.GetSubscribers("bridge-port")
+	sviSubs := eventbus.EBus.GetSubscribers("svi")
+	lbSubs := eventbus.EBus.GetSubscribers("logical-bridge")
+	vrfSubs := eventbus.EBus.GetSubscribers("vrf")
```

Comment:
can we be sure that number of subscribers won't change in between getObjectTypesToReplay gatherObjectsAndSubsToReplay calls?

Reply:
do we have a way to prevent adding new subscribers when we think that all of them are created?

Reply:
I think the question here is "from whom do we expect to add new subscribers ?" During runtime I do not see any process adding new subscribers. I do not think there is any trap here that can cause problems. So who do we expect to add new subscribers ?